src/Network/GRPC/MQTT/Message/Packet.hs

```diff
@@ -221,21 +250,26 @@ makePacketSender limit publish message = do
 
   if fromIntegral (ByteString.length message) <= maxPayloadSize
     then do
-      -- Perform a single publish on the current thread in case that the 
+      -- Perform a single publish on the current thread in case that the
       -- @message@ length is less than the maximum payload size.
       let packet = Packet message 0 1
       publish (wireWrapPacket' packet)
     else do
       caps <- liftIO getNumCapabilities
       jobs <- newTVarIO 0
 
-      replicateConcurrently_ caps $ fix \next ->
-        atomically (takeJobId jobs) >>= \case
-          Nothing -> pure ()
-          Just jobid -> do
-            let packet = makePacket jobid
-            publish (wireWrapPacket' packet)
-            next
+      replicateConcurrently_ caps do
+        let bufferSize :: Int
+            bufferSize = fromIntegral limit + metaDataSize
+         in withEncodeBuffer bufferSize \ptr -> do
+              fix \next -> do
+                atomically (takeJobId jobs) >>= \case
+                  Nothing -> pure ()
+                  Just jobid -> do
+                    let packet = makePacket jobid
+                    serialized <- encodeBufferOffPacket ptr bufferSize packet
+                    publish serialized
```

Comment:
Does `publish` guarantee to discard all references to its argument `ByteString` before returning to its caller in the IO state thread?  Because the bytes in that `ByteString` may change after that point.

Reply:
Yes, all references to underlying `MutableByteArray#` are discarded by the time the state token produced by `publish` is received to continue the loop.

Reply:
The `publish` function is actually an argument to `makePacketSender`.  Could you add a haddock comment to that argument explaining this requirement?  Most `ByteString` arguments are not so limited in temporal scope.